### DISCUSSION TITLE
json format not correct

### DISCUSSION THREAD

**Discussion by oym1994** [nlohmann/json#2489] [Q&A] [answered]
Hi, I use  nlohmann /json to create a json file. I want to watch it in vscode using json viewer but I get the error " json format not correct" and can't see the content. What should I do next?
Thank you!

**Comment by nlohmann** [comment, marked as answer]
How does the JSON file look like? What is the exact error message?

**Reply by oym1994** [reply]
Hi, thanks for ur response.

Actually ur project is adopte in OpenVSLAM. In the last function of this file [map_database.cc](https://github.com/xdspacelab/openvslam/blob/master/src/openvslam/data/map_database.cc) , the reconstructed map is converted 
to json file 
`
void map_database::to_json(nlohmann::json& json_keyfrms, nlohmann::json& json_landmarks) {
    std::lock_guard<std::mutex> lock(mtx_map_access_);

    // Save each keyframe as json
    spdlog::info("encoding {} keyframes to store", keyframes_.size());
    std::map<std::string, nlohmann::json> keyfrms;
    for (const auto id_keyfrm : keyframes_) {
        const auto id = id_keyfrm.first;
        const auto keyfrm = id_keyfrm.second;
        assert(keyfrm);
        assert(id == keyfrm->id_);
        assert(!keyfrm->will_be_erased());
        keyfrm->graph_node_->update_connections();
        assert(!keyfrms.count(std::to_string(id)));
        keyfrms[std::to_string(id)] = keyfrm->to_json();
    }
    json_keyfrms = keyfrms;

    // Save each 3D point as json
    spdlog::info("encoding {} landmarks to store", landmarks_.size());
    std::map<std::string, nlohmann::json> landmarks;
    for (const auto id_lm : landmarks_) {
        const auto id = id_lm.first;
        const auto lm = id_lm.second;
        assert(lm);
        assert(id == lm->id_);
        assert(!lm->will_be_erased());
        lm->update_normal_and_depth();
        assert(!landmarks.count(std::to_string(id)));
        landmarks[std::to_string(id)] = lm->to_json();
    }
    json_landmarks = landmarks;
}
`

Whie opeing the generated json file by using vscode with json viewer, it gives the following error:
"The file is not displayed in the edit because it  is either binary or uses an unsupported text encoding."

Thank you!

**Reply by nlohmann** [reply]
How do you write `json_landmarks` to a file?

**Reply by oym1994** [reply]
Here is the code for converting landmarks and keyframes to json file
`
nlohmann::json keyframe::to_json() const {
    // extract landmark IDs
    std::vector<int> landmark_ids(landmarks_.size(), -1);
    for (unsigned int i = 0; i < landmark_ids.size(); ++i) {
        if (landmarks_.at(i) && !landmarks_.at(i)->will_be_erased()) {
            landmark_ids.at(i) = landmarks_.at(i)->id_;
        }
    }

    // extract spanning tree parent
    auto spanning_parent = graph_node_->get_spanning_parent();

    // extract spanning tree children
    const auto spanning_children = graph_node_->get_spanning_children();
    std::vector<int> spanning_child_ids;
    spanning_child_ids.reserve(spanning_children.size());
    for (const auto spanning_child : spanning_children) {
        spanning_child_ids.push_back(spanning_child->id_);
    }

    // extract loop edges
    const auto loop_edges = graph_node_->get_loop_edges();
    std::vector<int> loop_edge_ids;
    for (const auto loop_edge : loop_edges) {
        loop_edge_ids.push_back(loop_edge->id_);
    }

    return {{"src_frm_id", src_frm_id_},
            {"ts", timestamp_},
            {"cam", camera_->name_},
            {"depth_thr", depth_thr_},
            // camera pose
            {"rot_cw", convert_rotation_to_json(cam_pose_cw_.block<3, 3>(0, 0))},
            {"trans_cw", convert_translation_to_json(cam_pose_cw_.block<3, 1>(0, 3))},
            // features and observations
            {"n_keypts", num_keypts_},
            {"keypts", convert_keypoints_to_json(keypts_)},
            {"undists", convert_undistorted_to_json(undist_keypts_)},
            {"x_rights", stereo_x_right_},
            {"depths", depths_},
            {"descs", convert_descriptors_to_json(descriptors_)},
            {"lm_ids", landmark_ids},
            // orb scale information
            {"n_scale_levels", num_scale_levels_},
            {"scale_factor", scale_factor_},
            // graph information
            {"span_parent", spanning_parent ? spanning_parent->id_ : -1},
            {"span_children", spanning_child_ids},
            {"loop_edges", loop_edge_ids}};
}

nlohmann::json landmark::to_json() const {
    return {{"1st_keyfrm", first_keyfrm_id_},
            {"pos_w", {pos_w_(0), pos_w_(1), pos_w_(2)}},
            {"ref_keyfrm", ref_keyfrm_->id_},
            {"n_vis", num_observable_},
            {"n_fnd", num_observed_}};

}
`

The final wrting code is:
`
nlohmann::json keyfrms;
    nlohmann::json landmarks;
    map_db_->to_json(keyfrms, landmarks);

    nlohmann::json json{{"cameras", cameras},
                        {"keyframes", keyfrms},
                        {"landmarks", landmarks},
                        {"frame_next_id", static_cast<unsigned int>(data::frame::next_id_)},
                        {"keyframe_next_id", static_cast<unsigned int>(data::keyframe::next_id_)},
                        {"landmark_next_id", static_cast<unsigned int>(data::landmark::next_id_)}};

    std::ofstream ofs(path, std::ios::out | std::ios::binary);

    if (ofs.is_open()) {
        spdlog::info("save the MessagePack file of database to {}", path);
        const auto msgpack = nlohmann::json::to_msgpack(json);
        ofs.write(reinterpret_cast<const char*>(msgpack.data()), msgpack.size() * sizeof(uint8_t));
        ofs.close();
    }
    else {
        spdlog::critical("cannot create a file at {}", path);
    }
`

The saved json file can be read by the json program, but it can't be openned in vscode.
Thank you!

**Reply by nlohmann** [reply]
You seem to write a binary file in MessagePack format, not a JSON file. This could explain the error message.

**Reply by oym1994** [reply]
Got it! Thank you for your kind response!
